pkg/pdp/service/roots_add.go

```diff
@@ -132,41 +154,70 @@ func (p *PDPService) AddRoots(ctx context.Context, id uint64, request []types.Ro
+		}
+
+		// turn the uploaded roots into PieceCIDV1
+		providedPieceCidV1, err := asPieceCIDv1(addReq.Root.String())
 		if err != nil {
-			return common.Hash{}, fmt.Errorf("failed to generate RootCID: %v", err)
+			return common.Hash{}, fmt.Errorf("failed to generate PieceCIDV1 for request: %w", err)
 		}
+
 		// Compare the generated and provided CIDs.
-		if !addReq.Root.Equals(generatedCID) {
-			return common.Hash{}, fmt.Errorf("provided RootCID does not match generated RootCID: %s != %s", addReq.Root, generatedCID)
+		if !providedPieceCidV1.Equals(generatedCID) {
+			return common.Hash{}, fmt.Errorf("provided RootCID does not match generated RootCID: %s (v1 %s) != %s",
+				addReq.Root, providedPieceCidV1, generatedCID)
 		}
 	}
 
 	// Step 5: Prepare the Ethereum transaction data outside the DB transaction
 	// Obtain the ABI of the PDPVerifier contract
-	abiData, err := contract.PDPVerifierMetaData()
+	abiData, err := smartcontracts.PDPVerifierMetaData()
 	if err != nil {
 		return common.Hash{}, fmt.Errorf("failed to get abi data from PDPVerifierMetaData: %w", err)
 	}
 
-	// Prepare RootData array for Ethereum transaction
-	// Define a Struct that matches the Solidity RootData struct
-	type RootData struct {
-		Root struct {
-			Data []byte
-		}
-		RawSize *big.Int
+	// Prepare PieceData array for Ethereum transaction
+	// Define a Struct that matches the Solidity PieceData struct
+	type PieceData struct {
+		Data []byte // CID
 	}
 
-	var rootDataArray []RootData
+	var pieceDataArray []PieceData
 
 	for _, addRootReq := range request {
 		// Convert RootCID to bytes
 		rootCID := addRootReq.Root
 
+		_, rawSize, err := commcid.PieceCidV1FromV2(rootCID)
+		if err != nil {
+			return common.Hash{}, fmt.Errorf("invalid PieceCIDV2: %w", err)
+		}
+		height, _, err := commcid.PayloadSizeToV1TreeHeightAndPadding(rawSize)
+		if err != nil {
+			return common.Hash{}, fmt.Errorf("computing height and padding: %w", err)
+		}
+		if height > 50 {
+			return common.Hash{}, fmt.Errorf("invalid height: %d", height)
+		}
+
+		// REVIEW: I am mega unsure of these code changes, but they do "pass" with the PDP Verifier contract.
+		// TODO(forrest): since migrating to the new contract, rod states there is an error
+		// in the curio size logic, context here: https://github.com/filecoin-project/curio/issues/650
+		// in commit 91aff56959407ec83171ef73d48c51fed8afb4c7 of curio
+		// Filecoin Team hash stated:
+		// subpieces are broken on our branch (called `rename` in Curio, don’t use it if you’re relying on subpieces.
+		// In fact, PieceCIDv2 isn’t quite working on our branch with this being an outstanding problem that’s currently
+		// biting us: https://github.com/filecoin-project/curio/issues/650; the not-quite-comprehensive switch to
+		// PieceCIDv2 hasn’t been a success, but we don’t really want to be changing the db schema just for that so we
+		// compromised and deferred the proper job to mkv2.
```

Comment:
In the code below I believe I have addressed this issue - but again unsure. We are able to use sub-pieces w/ pieceCIDv2, proving is passing proof set running from this branch may be seen here: https://pdp.vxb.ai/calibration/proofsets/255